Objects/listobject.c

```diff
@@ -180,6 +180,22 @@ PyList_New(Py_ssize_t size)
     return (PyObject *) op;
 }
 
+static inline PyObject *
+list_new(Py_ssize_t size) {
+    PyListObject *op = (PyListObject *) PyList_New(0);
+    if (size == 0 || op == NULL) {
+        return (PyObject *) op;
+    }
+    op->ob_item = (PyObject **) PyMem_Malloc(size * sizeof(PyObject *));
+    if (op->ob_item == NULL) {
+        Py_DECREF(op);
+        return PyErr_NoMemory();
+    }
+    Py_SIZE(op) = size;
```

Comment:
I'm not sure about this line. I would feel more confortable if it's done by the caller to always keep the list object consistent. For example, the caller could only set the size once the list is fully initialized. Otherwise, Py_DECREF(op) can crash on a half initialized list.

Reply:
I partly agree with this, but also note that any subsequent list array filling will necessarily incref the items along the way, so after the item array allocation and while the list isn't fully filled yet, it will necessarily be in an inconsistent and unsafe state that cannot simply be cleaned up via `Py_DECREF(op)`. Thus the careful selection of list operations to which this is applied: only those that do not risk raising exceptions within this state transition.

Letting the caller do the size adjustment does not save us here. But I agree that it slightly reduces the chance of programmer errors. So, +1. And that means we should also have a comment at this function that states this contract.

Reply:
I have the same feeling here with @scoder . Putting the size setting on the caller's side doesn't save us. If there is any exception raising cases inside the content setting logic, it's still painful since you have to maintain the current size. And the new helper function doesn't do all metadata setting now. I have not better solution now but I'd suggest add this note to the comment of the new helper function no matter which way it is in.